framework/Optimizers/FiniteDifferenceGradientOptimizer.py

```diff
@@ -0,0 +1,136 @@
+  """
+    Finite Difference Gradient Optimizer
+    This class currently inherits from the SPSA (since most of the gradient based machinery is there).
+    TODO: Move the SPSA machinery here (and GradientBasedOptimizer) and make the SPSA just take care of the
+    random perturbation approach
+  """
+  def __init__(self):
+    """
+      Default Constructor
+    """
+    SPSA.__init__(self)
+
+  def localInputAndChecks(self, xmlNode):
+    """
+      Local method for additional reading.
+      @ In, xmlNode, xml.etree.ElementTree.Element, Xml element node
+      @ Out, None
+    """
+    SPSA.localInputAndChecks(self, xmlNode)
+    self.paramDict['pertSingleGrad'] = len(self.fullOptVars)
+    self.gradDict['pertNeeded']      = self.gradDict['numIterForAve'] * (self.paramDict['pertSingleGrad']+1)
+
+
+  def _getPerturbationDirection(self,perturbationIndex, traj):
+    """
+      This method is aimed to get the perturbation direction (i.e. in this case the random perturbation versor)
+      @ In, perturbationIndex, int, the perturbation index (stored in self.perturbationIndeces)
+      @ In, traj, int, the trajectory id
+      @ Out, direction, list, the versor for each optimization dimension
+    """
+    optVars = self.getOptVars(traj)
+    if len(optVars) == 1:
+      if self.currentDirection:
+        factor = np.sum(self.currentDirection)*-1.0
+      else:
+        factor = 1.0
+      direction = [factor]
+    else:
+      if perturbationIndex == self.perturbationIndeces[0]:
+        direction = np.zeros(len(self.getOptVars(traj))).tolist()
+        factor = 1.0
+        if self.currentDirection:
+          factor = np.sum(self.currentDirection)*-1.0
+        else:
+          factor = 1.0
+        direction[0] = factor
+      else:
+        index = self.currentDirection.index(1.0) if self.currentDirection.count(1.0) > 0 else self.currentDirection.index(-1.0)
+        direction = self.currentDirection
+        newIndex = 0 if index+1 == len(direction) else index+1
+        direction[newIndex],direction[index] = direction[index], 0.0
+    self.currentDirection = direction
+    return direction
+
+  def localEvaluateGradient(self, optVarsValues, traj, gradient = None):
+    """
+      Local method to evaluate gradient.
+      @ In, optVarsValues, dict, dictionary containing perturbed points.
+                                 optVarsValues should have the form {pertIndex: {varName: [varValue1 varValue2]}}
+                                 Therefore, each optVarsValues[pertIndex] should return a dict of variable values
+                                 that is sufficient for gradient evaluation for at least one variable
+                                 (depending on specific optimization algorithm)
+      @ In, traj, int, the trajectory id
+      @ In, gradient, dict, optional, dictionary containing gradient estimation by the caller.
+                                      gradient should have the form {varName: gradEstimation}
```

Comment:
I don't think this input gets used.

Reply:
Yes not here but I wanted to keep it in case we need it in the future... I will remove it and in case we need it back we will add it later